numpyro/contrib/control_flow/scan.py

```diff
@@ -0,0 +1,64 @@
+from jax import lax, random, tree_flatten
+import jax.numpy as np
+
+import numpyro
+from numpyro import handlers
+from numpyro.primitives import _PYRO_STACK, apply_stack
+
+
+def scan_wrapper(fn, init_value, xs, rng_key=None, param_map=None, substitute_fn=None):
+
+    def body_fn(wrapped_carry, wrapped_x):
+        rng_key, carry = wrapped_carry
+        site_values, x = wrapped_x
+
+        if rng_key is not None:
+            rng_key, subkey = random.split(rng_key)
+            seeded_fn = fn if subkey is None else handlers.seed(fn, subkey)
+        else:
+            seeded_fn = fn
+
+        with handlers.block():
+            traced_fn = handlers.trace(handlers.substitute(seeded_fn, param_map=site_values,
+                                                           substitute_fn=substitute_fn))
+            carry, y = traced_fn(carry, x)
+        # we return 3 informations: distribution, value, is_subtituted
+        site_values = {name: site["value"] for name, site in traced_fn.trace.items()}
+        site_dists = {name: site["fn"] for name, site in traced_fn.trace.items()}
+        return (rng_key, carry), (site_values, site_dists, y)
+
+    length = tree_flatten(xs)[0][0].shape[0]
+    param_map = {} if (param_map is None and substitute_fn is None) else param_map
+    # only keeps sites have the same leading dimension as xs:
+    if param_map is not None:
+        param_map = param_map.copy()
+        for site_name in list(param_map.keys()):
+            if np.ndim(param_map[site_name]) == 0 or np.shape(param_map[site_name])[0] != length:
+                param_map.pop(site_name)
+    return lax.scan(body_fn, (rng_key, init_value), (param_map, xs))
+
+
+def scan(name, fn, init_value, xs, rng_key=None):
+    # if there are no active Messengers, we just run and return it as expected:
+    if not _PYRO_STACK:
+        (rng_key, carry), (site_values, site_dists, ys) = scan_wrapper(fn, init_value, xs, rng_key=rng_key)
```

Comment:
If we had a model where one of the sample sites was inside a scan body, and we ran inference and collected some samples, can we later use `handlers.condition` to condition the site to observed samples?

Reply:
We'll need to modify `condition` a bit, as in `substitute` to push `param_map` to `control_flow` primitive (probably to distinguish `substitute` and `condition`, we name it `condition_param_map` vs `substitute_param_map`). Then I think it should work. Later, we need to record more information from the trace such as `mask`, `scale`, `is_observed`... but for now, I think we can check if they are None or False and raise an error that asks for feature requests.

Reply:
I see, so currently this supports `trace`, `seed`, how about other effect handlers? It will be nice if whatever we have can transparently support additional effect handlers without causing any surprise. I will have to play with this though to clearly understand what are the issues involved.

Reply:
I think the point is how much information we want to collect from the `trace` (which we can return instead of returning distributions for the output of `scan_wrapper` as in the current implementation), which records more information. So far, `sample` primitive needs three more information `mask`, `scale`, `cond_indep_stack`. The first two ones are easily recorded. The last one is non-trivial but I guess we can resolve it by adding `cond_indep_stack` key to `control_flow` msg and based on `cond_indep_stack`, we'll wrap the scan body_fn by corresponding plate statements (before getting body_fn trace). Anyway, it is not important I guess (our predictive utilities using `vmap` instead of `plate` to draw multiple samples).